Userland/Services/AudioServer/Mixer.cpp

```diff
@@ -65,12 +67,19 @@ void Mixer::mix()
     decltype(m_pending_mixing) active_mix_queues;
 
     for (;;) {
-        if (active_mix_queues.is_empty() || m_added_queue) {
-            pthread_mutex_lock(&m_pending_mutex);
-            pthread_cond_wait(&m_pending_cond, &m_pending_mutex);
+        // While we have nothing to mix, wait on the condition.
+        if (m_pending_mixing.is_empty() && active_mix_queues.is_empty()) {
+            m_pending_mutex.lock();
+            m_mixing_necessary.wait_on(m_pending_mutex);
+            m_pending_mutex.unlock();
```

Comment:
No no no, this is _not_ how you use condition variables!!

I'm missing some context about how AudioServer works; but you must first take the lock, then do the check, and then sleep, and then while still holding the lock act on the new state, and only then release the lock. And the modifications that you wait for must only happen with the same lock held. Otherwise you get races or deadlocks.

Reply:
Here's how this typically looks using that helper function I suggested earlier:

```cpp
// Worker
mutex.lock();
while (true) {
    condvar.wait([&] {
        return !queue.is_empty();
    });
    for (auto item : queue)
        process(item);
}

// Producer
mutex.lock();
queue.push(item);
mutex.unlock();
condvar.signal();
```

Reply:
The testing doesn't reveal such deadlocks. In turn, when the old code was still doing the queue modifications while the lock was held after returning from wait(), it deadlocked every time. So I'll try something like your suggestion, but I'm not sure if it will work.

(I haven't really written any synchronization code before so :shrug: )

Reply:
I've done it properly now, I think, but you'll have to take another look. As for the lambda thing, I'd still love an explanation why that would be useful in my case / in general.

Reply:
> I haven't really written any synchronization code before

Ah, let me actually expand on why you have to do it like this and not in a different way.

See, a **condition variable** is not just a way to wait until someone notifies you. It "oversees" some _condition_ that can _repeatedly_ become true or false during the run time of the program (if you only need to handle a condition that can change from false to true _once_, you're better off using a simpler "one-shot event" primitive).

The most typical use case (and the one we seem to have here as well) of such a condition is whether some queue (aka channel, socket, pipe) is not empty (contains something). You usually have a _producer_ (aka writer) and a _consumer_ (reader), and there can be multiple of both kinds. So without condition variables, with some simple "notification" primitive, you'd write something like this:

```c++
SomeQueue queue;
Notification queue_not_empty_notification;

void producer(SomeType item)
{
    queue.enqueue(item);
    queue_not_empty_notification.notify();
}

void consumer()
{
    while (true) {
        if (queue.is_empty())
            queue_not_empty_notification.wait();
        SomeType item = queue.dequeue();
        process(item);
    }
}
```

The obviously wrong this with this code is the lack of synchronization. It's Undefined Behavior to access the same data (the queue) for both reading and writing concurrently (non-atomically).

So let's add a mutex. Importantly, modifications _and_ checking of the condition must be done while holding the mutex (it's instant UB otherwise):

```c++
Mutex mutex;
SomeQueue queue;
Notification queue_not_empty_notification;

void producer(SomeType item)
{
    mutex.lock();
    queue.enqueue(item);
    mutex.unlock();

    queue_not_empty_notification.notify();
}

void consumer()
{
    mutex.lock();
    while (true) {
        if (queue.is_empty()) {
            // Release the mutex while waiting, to let
            // the producer(s) enqueue new items.
            mutex.unlock();
            queue_not_empty_notification.wait();
            mutex.lock();
        }
        SomeType item = queue.dequeue();
        process(item);
    }
}
```

Makes sense so far?

This code *almost* gets the job done. The remaining issue is _lost wakeups_. Namely, if a producer manages to grab the mutex, enqueue an item, release the mutex, and signal the notification — all **in between** the consumer releasing the mutex and starting to wait, then the wake-up signaled by the producer will be lost and the consumer would be left waiting forever for the notification that will never come.

So _that's_ the issue condition variables were designed to resolve. A condition variable works a lot like the "notification" primitive we've used above, except that it can magically do unlock-and-wait _atomically_. (Of course, there's no actual magic, just careful use of futexes.) This way no producer can sneak in between the unlock and wait calls: we atomically unlock the mutex and start waiting, there's no in-between state where we have already unlocked the mutex, but are not waiting yet.

This is what the code looks like with condition variables:

```c++
Mutex mutex;
SomeQueue queue;
CondVar queue_not_empty_condvar;

void producer(SomeType item)
{
    mutex.lock();
    queue.enqueue(item);
    mutex.unlock();

    queue_not_empty_condvar.notify();
}

void consumer()
{
    mutex.lock();
    while (true) {
        while (queue.is_empty())
            queue_not_empty_condvar.wait();
        SomeType item = queue.dequeue();
        process(item);
    }
}
```

The single `queue_not_empty_condvar.wait()` call replaces the three calls from the previous listing:

```c++
mutex.unlock();
queue_not_empty_notification.wait();
mutex.lock();
```

That is to say, `CondVar::wait()` unlocks the mutex, waits until notified, then locks the mutex back, so upon returning you own the mutex again. Unlike the notification primitive, it does unlock-then-wait atomically.

Condition variables also have another trick up their sleeve: they avoid "thundering hurd" in case of broadcast notifications. Suppose you have multiple worker/consumer threads (forming a thread pool), and you want them to process the items in parallel. Then you'd release the mutex while processing each item, to let other threads in:

```c++
Mutex mutex;
SomeQueue queue;
CondVar queue_not_empty_condvar;

void producer(SomeType item)
{
    mutex.lock();
    queue.enqueue(item);
    mutex.unlock();

    queue_not_empty_condvar.notify_all();
}

void consumer()
{
    while (true) {
        mutex.lock();
        while (queue.is_empty())
            queue_not_empty_condvar.wait();
        SomeType item = queue.dequeue();
        mutex.unlock();
        process(item);
    }
}
```

In a poorly written condition variable implementation, after the `notify_all()` call all the waiting worker threads would wake up and contend for the mutex; of course only one thread will get it; this is what's called the "thundering hurd" problem. In a properly written condition variable implementation, the `notify_all()` will only really wake one thread, but ensure that it will wake the next thread when it releases the mutex (once it dequeues an item), and so on, the threads will end up waking each other in a chain. But this is "just" an internal optimization, it doesn't affect the semantics of mutexes / condition variables.

To summarize:
* Condition variable is about a _condition_
* The condition must be modified and checked only when holding a specific mutex, to do otherwise is UB
* Compared to a simpler notification primitive, condition variables help you avoid lost wakeups by providing an atomic unlock-and-wait operation

I hope this clarifies things a bit :^)

Reply:
It makes sense now, also why the lambda is such a good idea. I recommend you make a blog post out of this text, it's too good of an explanation to be wasted on a ephermal github review comment :^). I'll have another shot and be back with a new version in a couple hours.